json/src/main/scala/fs2/data/json/jq/TaggedMatcher.scala

```diff
@@ -0,0 +1,194 @@
+            dnf1.flatMap(conj1 => dnf2.map(conj2 => conj1.concatNel(conj2)))
+        }
+      case TaggedMatcher.AnyOf(clauses) =>
+        NonEmptyList
+          .fromList(clauses.toList)
+          .getOrElse(NonEmptyList.one(TaggedMatcher.Fail))
+          .flatMap(_.dnf)
+      case atom: AtomTaggedMatcher =>
+        NonEmptyList.one(NonEmptyList.one(atom))
+    }
+}
+private sealed trait AtomTaggedMatcher extends TaggedMatcher
+private sealed trait PatternTaggedMatcher extends AtomTaggedMatcher
+private sealed trait GuardTaggedMatcher extends AtomTaggedMatcher
+private sealed trait NegatableTaggedMatcher extends AtomTaggedMatcher
+private object TaggedMatcher {
+
+  case object StartJson extends PatternTaggedMatcher with NegatableTaggedMatcher
+  case object StartObject extends PatternTaggedMatcher with NegatableTaggedMatcher
+  case object StartArray extends PatternTaggedMatcher with NegatableTaggedMatcher
+
+  case class Field(name: String) extends PatternTaggedMatcher with NegatableTaggedMatcher
+  case class Index(idx: Int) extends PatternTaggedMatcher with NegatableTaggedMatcher
+  case class Slice(start: Int, end: Option[Int]) extends GuardTaggedMatcher with NegatableTaggedMatcher
+  case object Any extends PatternTaggedMatcher
+  case object Fail extends AtomTaggedMatcher
+
+  case class AnyOf(m: Set[TaggedMatcher]) extends TaggedMatcher
+  case class AllOf(m: Set[TaggedMatcher]) extends TaggedMatcher
+  case class Not(p: NegatableTaggedMatcher) extends GuardTaggedMatcher
+
+  implicit object TaggedInstances extends Pred[TaggedMatcher, TaggedJson] with Candidate[TaggedMatcher, TaggedJson] {
+
+    override def pick(set: TaggedMatcher): Option[TaggedJson] =
+      set match {
+        case StartJson            => Some(TaggedJson.StartJson)
+        case StartObject          => Some(TaggedJson.Raw(Token.StartObject))
+        case StartArray           => Some(TaggedJson.Raw(Token.StartArray))
+        case Field(name)          => Some(TaggedJson.StartObjectValue(name))
+        case Index(idx)           => Some(TaggedJson.StartArrayElement(idx))
+        case Slice(start, _)      => Some(TaggedJson.StartArrayElement(start))
+        case Any                  => Some(TaggedJson.Raw(Token.StartObject))
+        case Fail                 => None
+        case AnyOf(m)             => m.headOption.flatMap(pick(_))
+        case AllOf(m)             => m.headOption.flatMap(pick(_))
+        case Not(StartJson)       => Some(TaggedJson.Raw(Token.StartArray))
+        case Not(StartObject)     => Some(TaggedJson.Raw(Token.StartArray))
+        case Not(StartArray)      => Some(TaggedJson.Raw(Token.StartObject))
+        case Not(Field(name))     => Some(TaggedJson.StartObjectValue(s"!$name"))
+        case Not(Index(idx))      => Some(TaggedJson.StartArrayElement(idx + 1))
+        case Not(Slice(start, _)) => Some(TaggedJson.StartArrayElement(start - 1))
+      }
+
+    override def satsifies(p: TaggedMatcher)(e: TaggedJson): Boolean =
```

Comment:
```suggestion
    override def satisfies(p: TaggedMatcher)(e: TaggedJson): Boolean =
```

and of course call sites.

Reply:
🤦